### PR TITLE
feat: [CDS-50887]: Asg Blue Green Steps

### PR DESCRIPTION
**Pull request by vikrant-harness**
### Summary

- ASG Blue Green Rollback Step 
- ASG Blue Green Swap Service Step

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by vikrant-harness** [comment]
retrigger titlecheck

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  onChange?: (data: AsgSwapServiceData) => void
+  allowableTypes: AllowedTypes
+  stepViewType?: StepViewType
+  isNewStep?: boolean
+  template?: AsgSwapServiceData
+  readonly?: boolean
+  path?: string
+}
+
+function AsgSwapServiceWidget(
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx:60]: Can you create new component for this ? One component - one file please.

File: src/modules/75-cd/components/PipelineSteps/AsgBlueGreenRollbackStep/__tests__/__snapshots__/AsgBlueGreenRollback.test.tsx.snap [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Asg Blue Green Rollback Step tests Variables view renders fine 1`] = `<div />`;
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgBlueGreenRollbackStep/__tests__/__snapshots__/AsgBlueGreenRollback.test.tsx.snap:3]: This is a generic issue with all the steps using this generic component (Empty Spec) ,  Still looking for a fix for this.

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx (hunk 1 above)
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx]: Done

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+              </div>
+            </>
+          )
+        }}
+      </Formik>
+    </>
+  )
+}
+
+const AsgSwapServiceInputStep: React.FC<AsgSwapServiceProps> = ({
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx]: Done

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  validateInputSet({
+    data,
+    template,
+    getString,
+    viewType
+  }: ValidateInputSetProps<AsgSwapServiceData>): FormikErrors<AsgSwapServiceData> {
+    const isRequired = viewType === StepViewType.DeploymentForm || viewType === StepViewType.TriggerForm
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const errors = { spec: {} } as any
+    if (getMultiTypeFromValue(template?.timeout) === MultiTypeInputType.RUNTIME) {
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx]: Done

**Review by chetanvekariya-harness — DISMISSED** [review]

File: src/modules/75-cd/components/PipelineSteps/AsgBlueGreenRollbackStep/__tests__/__snapshots__/AsgBlueGreenRollback.test.tsx.snap (hunk 2 above)
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgBlueGreenRollbackStep/__tests__/__snapshots__/AsgBlueGreenRollback.test.tsx.snap:3]: No these do not come true in this case when spec is empty , In another places we pass variablesData and these appear twice in the drawer hence this comes true but we need to pass variablesData.spec

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import { PipelineStep } from '@pipeline/components/PipelineSteps/PipelineStep'
+import type { StringsMap } from 'stringTypes'
+import { AsgSwapServiceInputStep } from './AsgSwapServiceInputSet'
+import { AsgSwapServiceWidget } from './AsgSwapServiceEditWidget'
+import { validateGenericFields } from '../Common/GenericExecutionStep/utils'
+import pipelineVariablesCss from '@pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.module.scss'
+
+export interface AsgSwapServiceData extends StepElementConfig {
+  spec: AsgBlueGreenSwapServiceStepInfo
+  identifier: string
```
Comment (PratyushGar) [inline src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx:27]: Will we have seperate Swap routes for BG, canary, rolling, can we also ask BE to update type name for consistency

**Review by PratyushGar — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    <VariablesListTable
+      data={variablesData.spec}
+      originalData={initialValues.spec}
+      metadataMap={metadataMap}
+      className={pipelineVariablesCss.variablePaddingL3}
+    />
+  )
+}
+
+const AsgSwapServiceWidgetWithRef = React.forwardRef(AsgSwapServiceWidget)
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx:64]: This should have been done in the dedicated component..

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx (hunk 6 above)
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx:64]: Sure will take this up in subsequent PR

File: src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx (hunk 5 above)
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgSwapServiceStep/AsgSwapServiceStep.tsx:27]: No we will have Swap only for BlueGreen  , Will ask BE to change this in upcoming pr

**Pull request merged**